core/src/Kokkos_Serial.hpp

```diff
@@ -65,13 +67,40 @@
 #include <impl/Kokkos_FunctorAdapter.hpp>
 #include <impl/Kokkos_Tools.hpp>
 #include <impl/Kokkos_ExecSpaceInitializer.hpp>
+#include <impl/Kokkos_HostSharedPtr.hpp>
 
 #include <KokkosExp_MDRangePolicy.hpp>
 
 #include <Kokkos_UniqueToken.hpp>
 
 namespace Kokkos {
 
+namespace Impl {
+class SerialInternal {
+ public:
+  SerialInternal() = default;
+
+  bool is_initialized();
+
+  void initialize();
+
+  void finalize();
+
+  static SerialInternal& singleton();
+
+  std::mutex m_thread_team_data_mutex;
```

Comment:
Doesn't this need to be `static` if you want multiple copies of this class using the same resources?
```suggestion
  static std::mutex m_thread_team_data_mutex;
```

Reply:
There are no copy. We only access this class through the singleton.

Reply:
I think we should make sure that this class is not copy-constructible or copy-assignable then.

Reply:
`std::mutex` is not copy-constructible or copy-assignable. So the copy-constructor and the assignment operator are already deleted.